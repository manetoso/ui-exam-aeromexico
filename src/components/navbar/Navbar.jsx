import BookmarkIcon from '../../assets/icons/bookmark-fill-white.svg';
import AddIcon from '../../assets/icons/add-character-fill.svg';

export const Navbar = () => {
  return (
    <nav className='nav'>
      <div className='nav-container'>
        <div className='nav-wrapper'>
          <div className='nav-buttons-wrapper'>
            <button className='nav-button'>
              <span className='nav-button-label'>favoritos</span>
              <img
                className='nav-button-icon'
                src={BookmarkIcon}
                alt='bookmark icon'
              />
            </button>
            <div className='separator' />
            <button className='nav-button'>
              <span className='nav-button-label'>agregar</span>
              <img className='nav-button-icon' src={AddIcon} alt='add icon' />
            </button>
          </div>
        </div>
      </div>
    </nav>
  );
};
