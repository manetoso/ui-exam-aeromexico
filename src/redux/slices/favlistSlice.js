import { createSlice } from '@reduxjs/toolkit'

// REDUX SLICE FOR THE FAVLIST
export const favlistSlice = createSlice({
  // SLICE NAME
  name: 'favlist',
  // INITIAL STATE
  initialState: {
    list: []
  },
  // REDUCERS
  reducers: {
    // METHOD TO ADD IT TO THE FAVLIST
    addToFavlist: (state, action) => {
      // IF THERE ARE LESS THAN 5 ITEMS, ADD IT
      // OTHERWISE DON'T DO NOTHING
      state.list.length < 5 && state.list.push(action.payload)
    },
    // METHOD TO REMOVE IT FROM THE FAVLIST
    removeFromFavlist: (state, action) => {
      // SEARCH THE ID OF THE ITEM TO REMOVE
      // IF EXISTS REMOVE IT FROM THE ARRAY
      state.list = state.list.filter(item => item.id !== action.payload)
    }
  },
})

// EXPORTED ACTIONS (REDUCERS)
export const { addToFavlist, removeFromFavlist } = favlistSlice.actions
// EXPORTED MAIN REDUCER
export const favlistReducer = favlistSlice.reducer