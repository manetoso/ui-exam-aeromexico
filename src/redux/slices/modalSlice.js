import { createSlice } from '@reduxjs/toolkit'

// REDUX SLICE FOR THE MODAL
export const modalSlice = createSlice({
  // SLICE NAME
  name: 'modal',
  // INITIAL STATE
  initialState: {
    isOpen: false
  },
  // REDUCERS
  reducers: {
    // MODAL TOGGLER
    toggleModal: (state, action) => {
      state.isOpen = !state.isOpen
    },
  },
})

// EXPORTED ACTIONS (REDUCERS)
export const { toggleModal } = modalSlice.actions
// EXPORTED MAIN REDUCER
export const modalReducer = modalSlice.reducer