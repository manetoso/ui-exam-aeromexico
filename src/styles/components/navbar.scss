.nav {
  position: fixed;
  top: 0;
  width: 100%;
  transform: translateY(-200%);
  animation: nav-slice-in 500ms ease 500ms forwards;

  @media (max-width: $bp-phone) {
    bottom: 0;
    top: auto;
    transform: translateY(200%);
  }
}
.nav-container {
  display: flex;
  justify-content: center;
}
.nav-wrapper {
  display: flex;
  justify-content: flex-end;
  width: 1030px;

  @media (max-width: $bp-phone) {
    justify-content: center;
  }
}
.nav-buttons-wrapper {
  display: flex;
  position: relative;
}
.nav-button {
  @extend %title-14;

  align-items: center;
  background-color: $color-purple;
  border: 1px solid transparent;
  box-shadow: 0 4px 4px rgba($color: #000000, $alpha: 0.25);
  color: white;
  cursor: pointer;
  display: flex;
  gap: 0.3em;
  letter-spacing: 0.2em;
  outline: 3px solid transparent;
  text-transform: uppercase;
  transition: ease-in-out 150ms;
  padding-inline: 1em;
  padding-block: 0.25em;
  z-index: 99;

  &:focus {
    outline-color: $color-normal;
  }
  &:hover {
    transform: scale(1.03);
  }
  &:active {
    background-color: $color-purple;
    transform: scale(0.99);
  }

  &:first-child {
    border-bottom-left-radius: 5px;
  }
  &:last-child {
    border-bottom-right-radius: 5px;
  }

  @media (max-width: $bp-phone) {
    box-shadow: 0 -4px 4px rgba($color: #000000, $alpha: 0.25);
    padding-inline: 1.5em;
    padding-block: 0.25em;

    &:first-child {
      border-bottom-left-radius: 0px;
      border-top-left-radius: 5px;
    }
    &:last-child {
      border-bottom-right-radius: 0px;
      border-top-right-radius: 5px;
    }
  }
}
.nav-button-label {
}
.nav-button-icon {
}
.separator {
  background-color: $color-black;
  height: 100%;
  width: 1px;
  z-index: 99;
}

@keyframes nav-slice-in {
  100% {
    transform: translateY(0em);
  }
}
