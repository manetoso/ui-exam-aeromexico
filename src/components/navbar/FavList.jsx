import { useDispatch, useSelector } from 'react-redux';
import { removeFromFavlist } from '../../redux/slices/favlistSlice';
import Trash from '../../assets/icons/trash.svg';
import Placeholder from '../../assets/image-placeholder.svg';

export const FavList = ({ isOpen, toggleIsOpen }) => {
  // INITIALIZATION OF useSelector
  // PULLS THE DESIRE DATA FROM STORE => REDUCER => ELEMENT
  const { list } = useSelector(store => store.favlist);
  // INITILIZATION OF useDispatch
  // LET US EXECUTE THE ACTIONS DECLAIRE ON THE SLICE
  const dispatch = useDispatch();

  // METHOD TO REMOVE ELEMENT FROM FAVLIST
  // REQUIRES AN ID
  const handleRemoveFromFavlist = id => {
    dispatch(removeFromFavlist(id));
  };
  return (
    <>
      <div className={`favlist ${isOpen ? 'show-fav' : 'hidde-fav'}`}>
        {list.length === 0 && (
          <div className='fav-item'>
            <img className='item-image' src={Placeholder} alt='character' />
            <h3 className='item-name'>Lista Vacia</h3>
          </div>
        )}
        {list.length !== 0 &&
          list.map(({ id, name, image }) => (
            <div key={id} className='fav-item'>
              <img className='item-image' src={image} alt='character' />
              <h3 className='item-name'>{name}</h3>
              <button
                className='item-button'
                onClick={() => handleRemoveFromFavlist(id)}>
                <img src={Trash} alt='Delete icon' />
              </button>
            </div>
          ))}
      </div>
      {/* BACKGROUD TO CLOSE LIST WHEN CLICKING OUTSIDE */}
      <div
        className={`favlist-bg ${isOpen ? 'show-bg' : 'hidde-bg'}`}
        onClick={toggleIsOpen}
      />
    </>
  );
};
