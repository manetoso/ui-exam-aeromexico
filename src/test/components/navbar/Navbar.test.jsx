import { fireEvent, render, screen } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';

import { Provider } from 'react-redux';
import { favlistReducer } from '../../../redux/slices/favlistSlice';
import { modalReducer } from '../../../redux/slices/modalSlice';

import { Navbar } from '../../../components/navbar/Navbar';

/**
 * TESTS FOR THE Navbar.jsx COMPONENT
 */
describe('Testing in Navbar.jsx', () => {
  // STORE CONFIGURATION - JEST NEED IT
  const store = configureStore({
    reducer: {
      favlist: favlistReducer,
      modal: modalReducer,
    },
  });

  test('should match the snapshot', () => {
    // SHOULD MATCH THE SNAPSHOT SO THE RENDER MUST BE A CARD
    const { container } = render(
      <Provider store={store}>
        <Navbar />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  test('should have 2 buttons, fav button and modal toggler', () => {
    // RENDER
    render(
      <Provider store={store}>
        <Navbar />
      </Provider>
    );

    /**
     * SHOULD HAVE 2 BUTTONS
     * 	1 FOR THE FAVORITES LIST TOGGLE
     * 	1 FOR THE MODAL TOGGLER
     */
    expect(screen.getByText('favoritos')).toBeTruthy();
    expect(screen.getByText('agregar')).toBeTruthy();
  });

  test('should open the fav list when clicking on the right button', () => {
    // RENDER
    render(
      <Provider store={store}>
        <Navbar />
      </Provider>
    );

    /**
     * SHOULD OPEN THE FAV LIST
     * 	WE CAN KNOW IT BECAUSE THE CLASSNAME OF THE LIST
     * 	CHANGES FROM:
     * 	hidde-fav => show-fav
     */
    const btn = screen.getByText('favoritos');

    fireEvent.click(btn);
    const favlist = document.querySelector('.show-fav');

    expect(favlist).toBeTruthy();
    // screen.debug();
  });
});
