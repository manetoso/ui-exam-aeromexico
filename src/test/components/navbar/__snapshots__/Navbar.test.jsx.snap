// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing in Navbar.jsx should match the snapshot 1`] = `
<div>
  <nav
    class="nav"
  >
    <div
      class="nav-container"
    >
      <div
        class="nav-wrapper"
      >
        <div
          class="nav-buttons-wrapper"
        >
          <button
            class="nav-button"
          >
            <span
              class="nav-button-label"
            >
              favoritos
            </span>
            <img
              alt="bookmark icon"
              class="nav-button-icon"
              src="bookmark-fill-white.svg"
            />
          </button>
          <div
            class="separator"
          />
          <button
            class="nav-button"
          >
            <span
              class="nav-button-label"
            >
              agregar
            </span>
            <img
              alt="add icon"
              class="nav-button-icon"
              src="add-character-fill.svg"
            />
          </button>
          <div
            class="favlist hidde-fav"
          >
            <div
              class="fav-item"
            >
              <img
                alt="character"
                class="item-image"
                src="image-placeholder.svg"
              />
              <h3
                class="item-name"
              >
                Lista Vacia
              </h3>
            </div>
          </div>
          <div
            class="favlist-bg hidde-bg"
          />
        </div>
      </div>
    </div>
  </nav>
</div>
`;
